# Nivell 1 (màxim 5 punts)

## Pregunta 1 (1 punt)

Executa el següent codi en una cel·la:

```java
for(byte b = -125; b >= -130 && b <= 0; b--)
    println(b)
```

In [3]:
//Escriu i executa el codi aquí
for(byte b = -125; b >= -130 && b <= 0; b--)
    println(b)

-125
-126
-127
-128


In [7]:
byte by = (byte)(-128);
print(by);

-128

**Respon**:

1. Per què no s'imprimeix el -129 i el -130?
2. Per què cal posar la condició b <= 0 ?

---
Resposta 1:

Escriu la teva resposta aquí i **raona la teva resposta**.

IMPORTANT: cal raonar la resposta per poder puntuar.

Perquè el tipus de dada ``byte`` té el rang [-128,127] de valors vàlids, per la qual cosa quan el bucle arriba a -129, en estar fora del rang, es fa un _casting_ des d'``int`` a ``byte ``, resultant en 127, i en ser el 127 > 0, no compleix la condició ``b <= 0`` i per tant no s'imprimeix.

El mateix passa amb el -130, del qual se'n fa un casting a 126.

---
Resposta 2:

Escriu la teva resposta aquí i **raona la teva resposta**.

IMPORTANT: cal raonar la resposta per poder puntuar.

El -130 es _casteja_ a 126. Si b = 126, es compleix la condició ``b >= -130`` i el bucle continuaria (sense la condició ``b <= 0``) en motiu del ``b--``, passant a 125, 124, 123 ... fins arribar de nou a -130, que es _castejaria_ de nou a 126 i el bucle continuaria de forma _infinita_. Cal posar la condició ``b <= 0`` per a evitar el bucle _infinit_: el bucle acabarà amb el primer _casteig_ del -129 a 127.

## Pregunta 2 (1 punt)

Executa el següent codi en una cel·la:

```java
String lang2 = "Java";
String lang1 = "Hello";
println(lang1 + lang2);
println(lang1 + lang2 == "HelloJava");
```

In [5]:
//Escriu i executa el codi aquí
String lang2 = "Java";
String lang1 = "Hello";
println(lang1 + lang2);
println(lang1 + lang2 == "HelloJava");

HelloJava
false


**Respon**: 

Per què la última línia de codi imprimeix false?

---
Resposta:

Escriu la teva resposta aquí i **raona la teva resposta**.

IMPORTANT: cal raonar la resposta per poder puntuar.

**Cal tenir en compte que l'operador ``==`` sempre compara si les posicions en memòria de dos objectes és la mateixa posició o si són diferents. No compara el contingut de l'objecte.**

Per a comparar els continguts de dos objecte de tipus String usarem el mètode ``equals()``.

Important:

- Els String literals en Java es creen en una zona de memòria anomenada _String pool_. Els String en aquesta memòria sempre es reutilitzen (no hi ha String amb continguts repetits).
- Quan es crea un String usant ``new``, sempre es crea l'String en una zona de memòria anomenada _heap_. Cada objecte a la _heap_ és un objecte diferent als altres independentment dels seus continguts.

Així doncs, un String **literal** genera un nou objecte de tipus String en la memòria _String pool_ si no hi existeix prèviament un String amb els mateixos caràcters. Si aquell String ja existeix en aquesta memòria, no el crea de nou, sino que reutilitza el que ja hi ha. Per exemple:

In [23]:
String cadena1 = "java"; //es crea a la memòria String pool
String cadena2 = "java"; //reutilitza l'String "java" que existeix a la String pool
print(cadena1 == cadena2); //tenen la mateixa adreça de memòria

true

Per altra banda, si enlloc d'utilitzar un literal assignem un String a una variable utilitzant ``new``, es crearà un **nou** objecte de tipus String a la memòria _heap_, independentment de si el seu contingut és equivalent o no a un String prèviament existent a qualsevol zona de memòria. Per exemple:

In [24]:
String literal = "javabeans"; //Es crea a la zona de memòria String pool
String nou = new String("javabeans"); //Es crea a la zona de memòria Heap
print(literal == nou); //tenen adreces de memòria diferents

false

Per la seva banda, la concatenació de dos String **preexistents** també crea un **nou** String a la memòria _heap_. Per exemple:

In [25]:
String ja = "ja"; //es crea a l'String pool
String va = "va"; //es crea a l'String pool
String java = ja + va; //es crea a la Heap
String cadena = "java"; //es crea a l'String pool
print(java == cadena); //tenen adreces de memòria diferents

false

És a dir, ``String java = ja + va`` és equivalent a ``String java = new String(ja + va)``.

## Exercici 1 (1,5 punts)

Donada una cadena de text, a continuació teniu un snippet que canvia el primer caràcter pel segon:

In [14]:
String word = "java";

char[] letters = word.toCharArray();
char letter1 = letters[0];
char letter2 = letters[1];
letters[0] = letter2;
letters[1] = letter1;
print(new String(letters));


ajva

Aprofitant el codi anterior, feu un snippet que trobi totes les combinacions possibles que s'obtenen en intercanviar la primera lletra per cadascuna de les lletres restants.

Per exemple, per a "java" obtindrem:

```
ajva vaja aavj
```

El vostre codi ha de ser vàlid per a paraules de qualsevol mida.

In [33]:
//El vostre codi a continuació:
String word = "java";
for(int i = 0; i < word.length(); i++){
    char[] letters = word.toCharArray();
    char letter1 = letters[0];
    char letter2 = letters[i];
    letters[0] = letter2;
    letters[i]= letter1;
    printf("%s ",new String(letters));
}


java ajva vaja aavj 

## Exercici 2 (1,5 punts)

In [34]:
// Codi per a obtenir un array d'enters aleatoris

int getRandomNumber(int min, int max) {
    int upper = max * 10;
    Random random = new Random();
    return random.nextInt(upper - min) + min;
}

int[] createRandomIntArray(int max){
    int[] numbers = new int[max];
    for(int i = 0; i < max; i++){
        numbers[i] = getRandomNumber(0,max);
    }
    return numbers;
}

int[] numbers = createRandomIntArray(100);

print(Arrays.toString(numbers))

[379, 170, 654, 438, 481, 568, 179, 669, 758, 367, 688, 644, 336, 867, 986, 997, 221, 75, 546, 906, 766, 692, 509, 682, 979, 599, 375, 324, 189, 551, 995, 732, 794, 123, 792, 78, 52, 951, 483, 985, 903, 102, 592, 573, 939, 473, 704, 929, 268, 966, 618, 171, 477, 466, 417, 400, 707, 664, 322, 552, 640, 664, 986, 643, 690, 195, 426, 489, 592, 608, 309, 868, 520, 774, 894, 830, 250, 540, 854, 493, 264, 955, 898, 650, 767, 689, 880, 787, 652, 763, 13, 97, 262, 386, 815, 832, 906, 849, 941, 869]

***

Feu un snippet que, donat un array de números aleatoris com el generat amb el codi anterior, calculi **la mitjana** (que pot ser decimal) que s'obté de tots els números **parells** que pertanyen a l'array.

 

In [39]:
int[] numbers = createRandomIntArray(100);

//TODO (solució)

int sumEven = 0;

int count = 0;

for(int i = 0; i < numbers.length; i++){
    if(numbers[i] % 2 == 0){
        sumEven += numbers[i];
        count++;
    }
}

float average = (float)sumEven/count;

printf("Average of evens = %.2f", average);

Average of evens = 467,31